### PR TITLE
Deployment template and Installation instructions

### PR DESCRIPTION
**Pull request by akshaysngupta**

### CODE REVIEW

**Comment by akshaysngupta** [comment]
https://github.com/Azure/application-gateway-kubernetes-ingress/issues/33

File: docs/example/helm-config.yaml [hunk 1]
```diff
@@ -1,32 +1,9 @@
...
-  resourceGroup: <resourcegroup-name>
-  name: <applicationgateway-name>
-
-################################################################################
-# Specify which kubernetes namespace the ingress controller will watch
-# Default value is "default"
-#
-# kubernetes:
-#   watchNamespace: <namespace>
+  subscriptionId: 86501df2-2fc5-4fc3-b193-d305572e416f
```
Comment (vramakrishnan) [inline docs/example/helm-config.yaml]: Is this change intentional with hardcoded ID ?

File: docs/install-new.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+Once, we have the created the resources in Azure, we need to deploy following pods on the cluster:
+1) `aad pod identity` - This controller uses user assigned identity and provide ARM access token to the controller.
+2) `application gateway ingress controller` - This controller communicates ingress related events to the Application Gateway resource.
+
+Steps:
+
+1) Get credentials for the newly created Azure Kubernetes Cluster. This will cache the credentials in kubeconfig and set the context.  
+    ```bash
+    az aks get-credentials --resource-group <rg> --name <aksClusterName>`
```
Comment (vramakrishnan) [inline docs/install-new.md]: Maybe we should give the resource-group name and aks-cluster name here (that is used in template ?)
or provide steps to tell use how to get these names for rg and aks-cluster for completeness.

File: docs/install-new.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    ```bash
+    kubectl create -f https://raw.githubusercontent.com/Azure/aad-pod-identity/master/deploy/infra/deployment.yaml`
+    ```
+
+3) Install [Helm](https://docs.microsoft.com/en-us/azure/aks/kubernetes-helm) and run the following to add `application-gateway-kubernetes-ingress` helm package:
+    ```bash
+    helm init
+    helm repo add application-gateway-kubernetes-ingress https://azure.github.io/application-gateway-kubernetes-ingress/helm/
+    helm repo update
+    ```
```
Comment (vramakrishnan) [inline docs/install-new.md]: @asridharan  Is it possible to combine required k8s steps 2,3,4 into a single helm chart ?

File: docs/install-new.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    <a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fakshaysngupta%2Fapplication-gateway-kubernetes-ingress%2Fmaster%2Fdeploy%2Fazuredeploy.json" target="_blank">
+        <img src="http://azuredeploy.net/deploybutton.png"/>
+    </a>
+    <a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2Fakshaysngupta%2Fapplication-gateway-kubernetes-ingress%2Fmaster%2Fdeploy%2Fazuredeploy.json" target="_blank">
+        <img src="http://armviz.io/visualizebutton.png"/>
+    </a>
+
+    After the deployment completes, you can look at the Output window for parameters needed in the following steps.
+
```
Comment (vramakrishnan) [inline docs/install-new.md]: I think we should give a prescriptive example here, specifically hightlighting the "parameters" in sample log output.  
This will help the reader (removes the guesswork and possible config errors IMO)

**Comment by akshaysngupta** [comment]
https://github.com/akshaysngupta/application-gateway-kubernetes-ingress/blob/deploymentTemplate/docs/install-new.md

File: docs/install-new.md (hunk 4 above)
Comment (vramakrishnan) [reply docs/install-new.md]: thanks for adding the output.

File: docs/install-new.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    #
+    # kubernetes:
+    #   watchNamespace: <namespace>
+
+    ################################################################################
+    # Specify the authentication with Azure Resource Manager
+    #
+    # Two authentication methods are available:
+    # - Option 1: AAD-Pod-Identity (https://github.com/Azure/aad-pod-identity)
+    armAuth:
```
[the remark this answers is not in the document]
Comment (vramakrishnan) [reply docs/install-new.md]: Agree, please specify the command to use. For: eg (name has to match exactly what is created by template"
 " az identity show -g  <resource-group>  -n aaa-identity" here to get the required params below.

Also mention that "id" in the output maps to "identityResourceID" in this yaml. It is not intuitive for first time users, otherwise.

File: docs/install-new.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    ```yaml
+    # This file contains the essential configs for the ingress controller helm chart
+
+    ################################################################################
+    # Specify which application gateway the ingress controller will manage
+    #
+    appgw:
+        subscriptionId: <subscriptionId>
+        resourceGroup: <rresourceGroupName>
+        name: <applicationGatewayName>
```
Comment (vramakrishnan) [inline docs/install-new.md]: Since we are using template, please fix the name to "appgw"

File: docs/install-new.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+4) Edit [helm-config.yaml](example/helm-config.yaml) and fill in the values for `appgw` and `armAuth`
+    ```yaml
+    # This file contains the essential configs for the ingress controller helm chart
+
+    ################################################################################
+    # Specify which application gateway the ingress controller will manage
+    #
+    appgw:
+        subscriptionId: <subscriptionId>
+        resourceGroup: <rresourceGroupName>
```
Comment (vramakrishnan) [inline docs/install-new.md]: rresourceGroupName -> resourceGroupName

File: docs/install-new.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# Table of Contents
+- [Deploying the infrastructure on Azure](#deploying-the-infrastructure-on-azure)
+- [Setting up Application Gateway Ingress Controller on AKS](#setting-up-application-gateway-ingress-controller-on-aks)
+
+## Deploying the infrastructure on Azure
+
+To create the pre-requisite Azure resources, You can use the following template. It creates:
```
Comment (vramakrishnan) [inline docs/install-new.md]: You -> you

File: docs/install-new.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+## Setting up Application Gateway Ingress Controller on AKS
+
+Once, we have the created the resources in Azure, we need to deploy following controllers on the cluster:
+1) `aad pod identity` - This controller uses user assigned identity and provide ARM access token to the controller.
+2) `application gateway ingress controller` - This controller communicates ingress related events to the Application Gateway resource.
+
+Steps:
+
+1) To configure kubectl to connect to the deployed Azure Kubernetes Cluster, follow these [instructions](https://docs.microsoft.com/en-us/azure/aks/kubernetes-walkthrough#connect-to-the-cluster).
+
```
Comment (vramakrishnan) [inline docs/install-new.md]: Would be good to add a note, users must ensure that the right resource group and aks cluster name is used to initialize the kubeconfig as part of kubectl setup.   The name "aksCluster' also should match.

```
 az aks get-credentials --resource-group <resource-group>  --name aksCluster
```
For instance, I forgot reinitialize the kubeconfig to the new resource-group I created today and it ended up deploying/updating app-pod-identity in older cluster.

**Comment by vramakrishnan** [comment]
@akshaysngupta Tried the revised simplified instructions with ARM templates.
Works great ! Thanks.

File: docs/install-new.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+1) Azure Virtual Network with 2 subnets.
+2) Azure Application Gateway v2.
+3) Azure Kubernetes Service cluster with required permission to deploy nodes in the Virtual Network.
+4) User Assigned Identity to initialize the aad-pod-identity service and ingress controller.
+5) Set required RBACs.
+
+
+Steps:
+
+1) Create a service principal that will be assigned to aks cluster in the template.
```
Comment (asridharan) [inline docs/install-new.md]: s/assigned to/assigned to the

File: docs/install-new.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    ```bash
+    az ad sp create-for-rbac --skip-assignment
+    az ad sp show --id <appId> --query "objectId"
+    ```
+    **Note the appId, password and objectId.**
+
+2) After the above, click to create a custom template deployment. Provide the appId for servicePrincipalClientId, password, objectId in the parameters.
```
Comment (asridharan) [inline docs/install-new.md]: s/After the above/After creating the service principal in the step above, ....

Also modify the last part of the sentence to be:
Provide the appId for servicePrincipalClientId, password and objectId in the parameters.

File: docs/install-new.md (hunk 3 above)
Comment (asridharan) [reply docs/install-new.md]: @vramakrishnan the `aad-pod-identity` project is trying to come up with its own Helm, so waiting for that issue to be resolved before we take a dependency on them.

File: docs/install-new.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (asridharan) [reply docs/install-new.md]: @akshaysngupta can you provide the `az` command that can be used to retrieve the AD MSI for providing to the `Helm` config?

File: docs/install-new.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    <a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fakshaysngupta%2Fapplication-gateway-kubernetes-ingress%2Fmaster%2Fdeploy%2Fazuredeploy.json" target="_blank">
```
Comment (asridharan) [inline docs/install-new.md]: @akshaysngupta the URIs point to resources sitting in your fork. Could you create a separate PR, to commit the resources for the templates to this repository and we can than update this PR, to point to content in this repository.
